sdk/communication/azure-communication-chat/azure/communication/chat/_shared/user_credential.py

```diff
@@ -41,46 +44,88 @@ def get_token(self, *scopes, **kwargs):  # pylint: disable=unused-argument
+                    self._wait_till_lock_owner_finishes_refreshing()
                 else:
                     should_this_thread_refresh = True
                     self._some_thread_refreshing = True
                     break
 
         if should_this_thread_refresh:
             try:
-                newtoken = self._token_refresher()  # pylint:disable=not-callable
-
+                new_token = self._token_refresher()
+                if not self._is_token_valid(new_token):
+                    raise ValueError(
+                        "The token returned from the token_refresher is expired.")
                 with self._lock:
-                    self._token = newtoken
+                    self._token = new_token
                     self._some_thread_refreshing = False
                     self._lock.notify_all()
             except:
                 with self._lock:
                     self._some_thread_refreshing = False
                     self._lock.notify_all()
-
                 raise
+        if self._proactive_refresh:
+            self._schedule_refresh()
         return self._token
 
-    def _wait_till_inprogress_thread_finish_refreshing(self):
+    def _schedule_refresh(self):
+        if self._timer is not None:
+            self._timer.cancel()
+
+        token_ttl = self._token.expires_on - get_current_utc_as_int()
+
+        if self._is_token_expiring_soon(self._token):
+            # Schedule the next refresh for when it reaches a certain percentage of the remaining lifetime.
+            timespan = token_ttl // 2
+        else:
+            # Schedule the next refresh for when it gets in to the soon-to-expire window.
+            timespan = token_ttl - timedelta(
+                minutes=self._DEFAULT_AUTOREFRESH_INTERVAL_MINUTES).total_seconds()
+        if timespan <= TIMEOUT_MAX:
+            self._timer = Timer(timespan, self._update_token_and_reschedule)
+            self._timer.start()
+
+    def _wait_till_lock_owner_finishes_refreshing(self):
         self._lock.release()
         self._lock.acquire()
 
-    def _token_expiring(self):
-        return self._token.expires_on - get_current_utc_as_int() <\
-            timedelta(minutes=self._ON_DEMAND_REFRESHING_INTERVAL_MINUTES).total_seconds()
-
-    def _is_currenttoken_valid(self):
-        return get_current_utc_as_int() < self._token.expires_on
+    def _is_token_expiring_soon(self, token):
+        if self._proactive_refresh:
+            interval = timedelta(
+                minutes=self._DEFAULT_AUTOREFRESH_INTERVAL_MINUTES)
+        else:
+            interval = timedelta(
+                minutes=self._ON_DEMAND_REFRESHING_INTERVAL_MINUTES)
+        return ((token.expires_on - get_current_utc_as_int())
+            < interval.total_seconds())
+
+    @classmethod
+    def _is_token_valid(cls, token):
+        return get_current_utc_as_int() < token.expires_on
+
+    def __enter__(self):
```

Comment:
If a customer _doesn't_ use a context manager - when's the first time the `schedule_refresh` will be called?
If I'm following the changes correctly - it looks like it will happen at the end of `get_token`, during the `_update_token_and_reschedule` call?

Reply:
Yeah, as you have mentioned in your previous [comment](https://github.com/Azure/azure-sdk-for-python/pull/21834#discussion_r819115398) for people not using a context manager the scheduling happens from the first API call in the _update_token_and_reschedule.

Reply:
Added [async](https://github.com/Azure/azure-sdk-for-python/pull/21834/commits/be1aa9f9e6e232bf443706daac6ce17d6c974674#diff-dbfc5aa9546a717c64a9b0df2b572ecdda52759bde3aa7e51af95c44a700be80) and [sync](https://github.com/Azure/azure-sdk-for-python/pull/21834/commits/be1aa9f9e6e232bf443706daac6ce17d6c974674#diff-616d1098edeb23af7081e6a0b40703cb5e2ca2b0b97b680fef43e99ec813eac1) tests for the using credential without a context manager scenarios